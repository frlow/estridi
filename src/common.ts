import { Mode } from './generators/index.js'
import { ProcessorType } from './processors/index.js'

export const allowedRegex = /[^a-zA-Z0-9åäöÅÄÖ ]/g
export type GenerationResult = { file: string, content: string, overwrite: boolean }
export type Scraped = {
  type: string
  inputs?: string[]
  text: string
  connections: Record<string, string>
  id: string
  actions?: string[]
  headers?: string[]
  content?: string[][]
  linked?: string
}[]
export const testedNodeTypes = ['message', 'script', 'subprocess', 'signalSend']
export const scrapedFile = 'scraped.json'
export type EstridiConfig = { mode: Mode, token: string, fileId: string, processorType?: ProcessorType }
