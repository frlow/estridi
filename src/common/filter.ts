export const filterScraped = (scraped: Scraped, rootName: string): Scraped => {
  const processNode = (node: any, distance: number) => {
    if (!node) return
    if (acc[node.id]) return
    node.distance = distance
    acc[node.id] = node
    if (node.next) processNode(scraped.find((n) => n.id === node.next), distance + 1)
    if (node.options)
      Object.keys(node.options).forEach((o) =>
        processNode(scraped.find((n) => n.id === o), distance + 1)
      )
    if (node.actions)
      Object.keys(node.actions).forEach((a) =>
        processNode(scraped.find((n) => n.id === a), distance + 1)
      )
    if (node.link) processNode(scraped.find((n) => n.id === node.link), distance + 1)
  }
  const acc: Record<string, ScrapedNode> = {}
  const root = scraped.find(
    (r: ScrapedStart) => r.isRoot && r.raw === rootName
  )
  processNode(root, 0)
  // const rootNodes = scraped.filter((node: ScrapedStart) => node.isRoot)
  // rootNodes.forEach(n => processNode(n))
  scraped.filter((s) => s.type === 'table').forEach((t) => (acc[t.id] = t))
  return Object.values(acc)
}
