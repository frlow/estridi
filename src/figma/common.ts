export const allowedRegex = /[^a-zA-Z0-9åäöÅÄÖ ]/g
