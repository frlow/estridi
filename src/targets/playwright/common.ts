import { _ } from '../../common/texts.js'
import { generateScriptTest, generateServiceCallTest } from './testScript.js'
import { generateSubprocessTableTest } from './subprocessTable.js'

export const rawCommentBlock = (raw: string) => `/*
${raw}
*/`.replace(/^/gm, _(2))

export const allowedRegexPermissive = /[^a-zA-Z0-9åäöÅÄÖ\[]\(\)-_ ]/g

export const getRootNote = (scraped: Scraped): string => {
  const root = scraped.find((n: ScrapedStart) => n.isRoot)
  if (!root?.note) return '\n\n'
  else return `/*
${root.note.replaceAll(allowedRegexPermissive, ' ')}
*/
`
}

export const generateTest = (
  scraped: Scraped,
  node: ScrapedNode,
  blockPath: any[],
  usedNames: Record<string, number>
): string => {
  if (node.type === 'script') {
    return generateScriptTest(scraped, node, blockPath, usedNames)
  } else if (node.type === 'serviceCall') {
    return generateServiceCallTest(scraped, node, blockPath, usedNames)
  } else if (node.type === 'subprocess' && node.tableKey) {
    return generateSubprocessTableTest(scraped, node, blockPath)
  } else {
    debugger
  }
}
