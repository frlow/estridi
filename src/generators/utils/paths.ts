import { Scraped } from '../../common'

export const findAllPaths = (scraped: Scraped) => {
  const root = scraped.find(node => node.type === 'start' && node.text === 'root')
  if (!root) {
    console.warn('Could not find root node!')
    return
  }
  const process = (start: string[]) => {
    const paths: string[][] = []
    const toProcess = [start]
    while (toProcess.length > 0) {
      const path = toProcess.pop()!
      const id = path[path.length - 1].replace('*', '')
      const current = scraped.find(node => node.id === id)
      const connections = Object.keys(current.connections || {})
      if (current.type === 'subprocess' && !path[path.length - 1].includes('*')) {
        const subprocess = scraped.find(node => node.type === 'start' && node.text === current.text)
        if (subprocess) {
          const subprocessProcessed = process([subprocess.id])
          subprocessProcessed.forEach(sp => toProcess.push([...path, ...sp, `*${current.id}`]))
          continue
        } else {
          if (!['Validate', 'Display'].some(t => current.text.startsWith(t)))
            console.warn(`Could not find subprocess: ${current.text}`)
        }
      }
      if (connections.length === 0) {
        const pathTemp = path.filter(id => !id.includes('*'))
        paths.push(pathTemp)

        // Locate the last user action taken, all actions before this must be filtered
        let lastSignalListen = pathTemp.length - 1
        for (let i = pathTemp.length - 1; i >= 0; i--) {
          if (scraped.find(node => node.id === pathTemp[i]).type === 'signalListen') break
          lastSignalListen = i
        }
        const signalListens = pathTemp
          .slice(lastSignalListen)
          .flatMap(id => scraped.find(node => node.id === id).actions || [])
        signalListens.forEach(sl => toProcess.push([...pathTemp, sl]))
      } else
        connections.forEach(c => toProcess.push([...path, c]))
    }
    return paths
  }
  const processed = process([root.id])
  console.log(processed.length)
}
