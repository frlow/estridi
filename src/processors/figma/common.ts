import { Scraped } from '../../common.js'

export const processNode = (node: any, acc: Record<string, any>) => {
  if (acc[node.id]) return
  acc[node.id] = node
  const children = node.children || []
  children.forEach((c: any) => processNode(c, acc))
}

export const processFigmaDocument = (document: any, getNodeMetadata: (node: any) => any): Scraped => {
  const acc: Record<string, any> = {}
  processNode(document, acc)
  const nodes = Object.values(acc)
  nodes.forEach(n => n.scraped = nodes)
  const mappedNodes = nodes.map(node => getNodeMetadata(node)).filter(n => n)
  mappedNodes.filter(node => node.type === 'subprocess').forEach(node => {
    const link = mappedNodes.find(n => n.type === 'start' && n.text === node.text)
    if (link) node.linked = link.id
  })
  return mappedNodes
}

export function getConnections(node: any) {
  return node.scraped
    .filter((n: any) =>
      n.type === 'CONNECTOR' &&
      !n.strokeDashes &&
      n.connectorStart.endpointNodeId === node.id)
    .reduce((acc: any, cur: any) => ({ ...acc, [cur.connectorEnd.endpointNodeId]: cur.name || 'N/A' }), {})
}

export const getTableMetadata = (node: any) => {
  if (node.type !== 'TABLE') return undefined
  const rows: string[][] = Object.values(node.children.reduce((acc: Record<string, string[]>, cur: any) => ({
    ...acc,
    [cur.absoluteBoundingBox.y]: [...(acc[cur.absoluteBoundingBox.y] || []), cur.characters]
  }), {}))
  const headers = rows[0]
  const corner = rows[0][0]
  if (!corner.startsWith('.')) return undefined
  const content = rows.slice(1)
  return {
    type: 'table',
    text: corner.substring(1),
    headers,
    content,
    id: node.id
  }
}

type Points = { x0: number, x1: number, y0: number, y1: number }
export const isNodeInside = (host: any, child: any) =>
  isInside({
    x0: host.x,
    x1: host.x + host.width,
    y0: host.y,
    y1: host.y + host.height
  }, {
    x0: child.x,
    x1: child.x + child.width,
    y0: child.y,
    y1: child.y + child.height
  })
export const isInside = (host: Points, child: Points) => {
  const compare = (x: number, y: number) => x > host.x0 && x < host.x1 && y > host.y0 && y < host.y1
  if (compare(child.x0, child.y0)) return true
  if (compare(child.x1, child.y0)) return true
  if (compare(child.x0, child.y1)) return true
  if (compare(child.x1, child.y1)) return true
  return false
}

export type NodeType =
  'script' |
  'serviceCall' |
  'userAction' |
  'start' |
  'end' |
  'gateway' |
  'subprocess' |
  'signalListen'

export type NodeMetadata = {
  type: NodeType
}
