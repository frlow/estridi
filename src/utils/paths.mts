import { Scraped } from '../common.js'

export const findAllPaths = (scraped: Scraped) => {
  const root = scraped.find(node => node.type === 'start' && node.text?.startsWith("root"))
  if (!root) {
    console.warn('Could not find root node!')
    return []
  }
  const process = (start: string) => {
    const paths: string[][] = []
    const toProcess: { nodes: string[], stack: string[], current: string }[] = [{
      nodes: [start],
      stack: [],
      current: start
    }]
    const missingSubProcesses: Record<string, null> = {}
    while (toProcess.length > 0) {
      const path = toProcess.pop()!
      const current = scraped.find(node => node.id === path.current)
      const connections: string[] = [...Object.keys(current.connections || {}), ...(current.actions || [])]
      if (current.type === 'subprocess') {
        if (connections.length > 1) {
          throw 'Not supported!'
        }
        const subProcess = scraped.find(s => s.text === current.text && s.type === 'start')
        if (!subProcess) {
          missingSubProcesses[current.text] = null
          toProcess.push(...connections.map((c) => ({
            nodes: [...path.nodes, c],
            current: c,
            stack: [...path.stack]
          })))
        } else {
          toProcess.push({
            nodes: [...path.nodes, subProcess.id],
            current: subProcess.id,
            stack: [...path.stack, ...connections]
          })
        }
      } else if (connections.length === 0) {
        if (path.stack.length > 0) {
          const returnId = path.stack[path.stack.length - 1]
          toProcess.push({
            nodes: [...path.nodes, returnId],
            current: returnId,
            stack: path.stack.slice(0, path.stack.length - 1)
          })
        } else paths.push(path.nodes)
      } else
        toProcess.push(...connections.map((c) => ({
          nodes: [...path.nodes, c],
          current: c,
          stack: [...path.stack]
        })))
    }
    console.warn(missingSubProcesses)
    return paths
  }
  return process(root.id)
}
