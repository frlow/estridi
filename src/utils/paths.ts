import { Scraped } from '../common.js'

const getCurrent = (item: ProcessItem, scraped: Scraped) => scraped.find(node => node.id === item.nodes[item.nodes.length - 1])
const getConnections: (current: any, nodes: string[]) => string[] = (current, nodes) =>
  ([...Object.keys(current.connections || {}), ...(current.actions || [])])
    .filter(n => !nodes.includes(n)) // Filtering out loops

type ProcessItem = { nodes: string[], stack: string[] }

export const findAllPaths = (scraped: Scraped, rootId: string) => {
  const finishedPaths: string[][] = []
  const toProcess: ProcessItem[] = [{
    nodes: [rootId],
    stack: []
  }]
  while (toProcess.length > 0) {
    const path = toProcess.pop()!
    const current = getCurrent(path, scraped)
    const connections = getConnections(current, path.nodes)
    const stack = [...path.stack]
    if (connections.length === 0 && path.stack.length === 0 && !current.linked) {
      finishedPaths.push(path.nodes)
      continue
    }
    if (connections.length === 0 && stack.length > 0) {
      const fromStack = stack.splice(stack.length - 1, 1)[0]
      connections.push(fromStack)
    }

    const nextItems = current.linked ?
      [{
        stack: [...stack, ...connections],
        nodes: [...path.nodes, current.linked]
      }] :
      connections.map(c =>
        ({
          stack: [...stack],
          nodes: [...path.nodes, c]
        })
      )

    toProcess.push(...nextItems)
  }
  return finishedPaths
}
