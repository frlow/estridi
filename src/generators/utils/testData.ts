import { Scraped } from '../../common'
import { findAllPaths } from './paths'

export const getTestData = (scraped: Scraped) => findAllPaths(scraped).flatMap(n => n).filter((n, i, arr) => arr.indexOf(n) === i).map(id => scraped.find(s => s.id === id))
