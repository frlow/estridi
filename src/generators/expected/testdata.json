[
  {
    "name": "540466 User Story: Select Account",
    "scenarios": [
      {
        "given": [
          {
            "type": "connector",
            "text": "Errors from demo",
            "value": "yes"
          }
        ],
        "when": [],
        "then": [
          {
            "type": "serviceCall",
            "text": "api call demo"
          },
          {
            "type": "message",
            "text": "Could not load page"
          }
        ]
      },
      {
        "given": [
          {
            "type": "connector",
            "text": "Errors from demo",
            "value": "no"
          },
          {
            "type": "connector",
            "text": "A or B",
            "value": "A"
          }
        ],
        "when": [],
        "then": [
          {
            "type": "serviceCall",
            "text": "api call demo"
          },
          {
            "type": "message",
            "text": "Display stuff"
          },
          {
            "type": "subprocess",
            "text": "Some subflow"
          },
          {
            "type": "subprocess",
            "text": "Validate fields Payment data"
          }
        ]
      },
      {
        "given": [
          {
            "type": "connector",
            "text": "Errors from demo",
            "value": "no"
          },
          {
            "type": "connector",
            "text": "A or B",
            "value": "B"
          }
        ],
        "when": [],
        "then": [
          {
            "type": "serviceCall",
            "text": "api call demo"
          },
          {
            "type": "message",
            "text": "Display stuff"
          },
          {
            "type": "script",
            "text": "Some script"
          },
          {
            "type": "subprocess",
            "text": "Validate fields Payment data"
          }
        ]
      },
      {
        "given": [
          {
            "type": "connector",
            "text": "Errors from demo",
            "value": "no"
          },
          {
            "type": "connector",
            "text": "A or B",
            "value": "B"
          }
        ],
        "when": [
          {
            "type": "signalListen",
            "text": "Take action"
          }
        ],
        "then": [
          {
            "type": "serviceCall",
            "text": "api call demo"
          },
          {
            "type": "message",
            "text": "Display stuff"
          },
          {
            "type": "script",
            "text": "Some script"
          },
          {
            "type": "script",
            "text": "Show something"
          },
          {
            "type": "subprocess",
            "text": "Saved payee Yet another subflow"
          }
        ]
      },
      {
        "given": [
          {
            "type": "connector",
            "text": "Errors from demo",
            "value": "no"
          },
          {
            "type": "connector",
            "text": "A or B",
            "value": "B"
          }
        ],
        "when": [
          {
            "type": "signalListen",
            "text": "Cancel"
          }
        ],
        "then": [
          {
            "type": "serviceCall",
            "text": "api call demo"
          },
          {
            "type": "message",
            "text": "Display stuff"
          },
          {
            "type": "script",
            "text": "Some script"
          },
          {
            "type": "script",
            "text": "Clear page"
          }
        ]
      }
    ],
    "tables": [
      {
        "header": [
          "Component type",
          "Properties",
          "Special Requirement"
        ],
        "rows": [
          {
            "label": "Land",
            "values": [
              "DropdownInputSelectDesktop",
              "RequiredString Format 12 characters",
              "Default value Sverige"
            ]
          },
          {
            "label": "Mottagarens namn",
            "values": [
              "Input field",
              "RequiredString Format 135 characters",
              ""
            ]
          },
          {
            "label": "Gatuadress",
            "values": [
              "Input field",
              "RequiredString Format 135 characters",
              ""
            ]
          },
          {
            "label": "Postnummer",
            "values": [
              "Input field",
              "Required String Format 135 characters",
              ""
            ]
          }
        ]
      }
    ]
  }
]
